crates/bevy_hierarchy/src/hierarchy.rs

```diff
@@ -20,6 +23,29 @@ pub struct DespawnChildrenRecursive {
     pub entity: Entity,
 }
 
+/// System to print a warning if at any time we detect an inconsistent hierarchy.
+pub fn hierarchy_healthcheck_system<T: Component>(
```

Comment:
This feels a bit like a "debug diagnostic" system. Not really useful in the context of a released app / worth paying the runtime cost. Maybe we should work out a pattern to enable these systems by default, but only during development time? (ex: only enable when compiled in debug mode).

Reply:
Strongly agree with this. I considered doing so. But there are many ways of doing this. Here are some:

1. Adding a field to the `TransformPlugin` and `RenderPlugin` the user can toggle
2. Add fields to the `DiagnosticPlugin` and check them when adding the heath check systems (bonus point for moving the health check systems to it)
3. Check log level when adding the system (would it be possible to check for `bevy_transform` and `bevy_render` specific log levels?)
4. Use `#[cfg(debug_assertion)]`
5. Some kind of feature flag

The hierarchy health check system also is pretty cheap due to the `Added<Parent>` query filter.

I'm now convinced that (2) is the right solution, but when making this PR I figured any decision made would be controversial, and I wanted to keep this PR focused in scope. Additionally, the UI node diagnostic cannot be toggled off (however, unlike this diagnostic, it doesn't require an additional system)